src/main/java/seedu/duke/data/TransactionList.java

```diff
@@ -188,11 +202,145 @@ public String findTransactions(String keywords) {
+     * @param date A specified date to backdate N months from occurring month.
+     * @return An array containing the start and end date.
+     */
+    public static LocalDate[] getMonthRange(LocalDate date, int numberOfMonths) {
+        // Solution below adapted from https://stackoverflow.com/a/51356522
+        LocalDate lastMonth = date.minusMonths(1);
+        LocalDate from = date.minusMonths(numberOfMonths).withDayOfMonth(1);
+        LocalDate to = lastMonth.withDayOfMonth(lastMonth.getMonth().maxLength());
+
+        return new LocalDate[]{from, to};
+    }
+
+    /**
+     * Gets all transactions recorded on a specific year.
+     *
+     * @param year  A specified year.
+     * @return An array list containing all transactions recorded on a specific year.
+     */
+    public ArrayList<Transaction> getTransactionsByYear(int year) {
+        //@@author chydarren-reused
+        //Reused from https://stackoverflow.com/a/69440139
+        // with minor modifications
+        ArrayList<Transaction> transactionsByYear = (ArrayList<Transaction>) transactions.stream()
+                .filter(transaction -> Year.from(transaction.getDate()).equals(Year.of(year)))
+                .collect(Collectors.toList());
+        //@@author
+
+        return transactionsByYear;
+    }
+
+    /**
+     * Gets all transactions recorded on a specific month.
+     *
+     * @param year  A specified year.
+     * @param month A specified month within the year.
+     * @return An array list containing all transactions recorded on a specific month.
+     */
+    public ArrayList<Transaction> getTransactionsByMonth(int year, int month) {
+        //@@author chydarren-reused
+        //Reused from https://stackoverflow.com/a/69440139
+        ArrayList<Transaction> transactionsByMonth = (ArrayList<Transaction>) transactions.stream()
+                .filter(transaction -> YearMonth.from(transaction.getDate()).equals(YearMonth.of(year, month)))
+                .collect(Collectors.toList());
+        //@@author
+
+        return transactionsByMonth;
+    }
+
+    /**
+     * Gets all transactions recorded on the last N weeks or months.
+     *
+     * @param numberOfType  An integer that represents the last N number of weeks or months.
+     * @param type          A string that represents the getting of N "weeks", or "months".
+     * @return An array list containing all transactions recorded on the last N number of weeks or months.
+     */
+    public ArrayList<Transaction> getTransactionsByLastN(int numberOfType, String type) {
+        ArrayList<Transaction> transactionsByLastN = new ArrayList<>();
+        LocalDate[] dateRange = {};
+
+        if (type.equals(WEEKS)) {
+            dateRange = getWeekRange(LocalDate.now(), numberOfType);
+        } else {
+            assert type.equals(MONTHS);
+            dateRange = getMonthRange(LocalDate.now(), numberOfType);
+        }
```

Comment:
The if-else here seems to rely on the usage of assert to ensure that no other input for "type" is accepted other than WEEKS or MONTHS. AFAIK, this is not the recommended way of using assert.

I would suggest changing this if-else into if-elseif-else instead, and to throw an error when the else part of the code is arrived at. This way, we do not rely on the assert to ensure that the type is only WEEKS or MONTHS.

Reply:
Ok sure! I agree with this change